File: src/storage_manager.py
#!/usr/bin/env python3
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright 2017 - Edoardo Morassutto <edoardo.morassutto@gmail.com>
# Copyright 2017 - Luca Versari <veluca93@gmail.com>

import os
import os.path
from .config import Config


class StorageManager:

    # TODO think about that
    ALLOWED_CHARSET = "[a-zA-Z0-9_-.()\[\]]"
    MAX_LENGTH = 100

    @staticmethod
    def new_source_file(source_id, filename):
        """
        Get the relative path to the source file with the specified filename
        :param source_id: id of the source file
        :param filename: name of the file, will be sanitized
        :return: A path relative to the Config.storedir where save that file
        """
        filename = StorageManager._sanitize(filename)
        path_prefix = [source_id[:2], source_id[2:4]]
        return os.path.join("source", *path_prefix, source_id, filename)

    @staticmethod
    def new_output_file(output_id, filename):
        """
        Get the relative path to the output file with the specified filename
        :param output_id: id of the output file
        :param filename: name of the file, will be sanitized
        :return: A path relative to the Config.storedir where save that file
        """
        filename = StorageManager._sanitize(filename)
        path_prefix = [output_id[:2], output_id[2:4]]
        return os.path.join("output", *path_prefix, output_id, filename)

    @staticmethod
    def new_input_file(input_id, task_name, attempt):
        """
        Get the relative path to the input file for the specified attempt
        on the specified task
        :param input_id: id of the input file
        :param task_name: name of the task
        :param attempt: attempt on the task
        :return: A path relative to the Config.storedir where that file is
        """
        filename = task_name + "_input_" + str(attempt) + ".txt"
        path_prefix = [input_id[:2], input_id[2:4]]
        filename = os.path.join("input", *path_prefix, input_id, filename)
        StorageManager._create_dir(StorageManager.get_absolute_path(filename))
        return filename

    @staticmethod
    def get_file_size(filename):
        """
        Get the size of the filename in bytes
        :param filename: Relative path of the file to check
        :return: The number of bytes of the file
        """
        absolute_path = StorageManager.get_absolute_path(filename)
        return os.stat(absolute_path).st_size

    @staticmethod
    def save_file(relative_path, file_content):
        """
        Store a file in the filesystem, creates the directories needed
        :param relative_path: Relative path of the file
        :param file_content: Content of the file
        """
        absolute_path = StorageManager.get_absolute_path(relative_path)
        StorageManager._create_dir(absolute_path)
        file = open(absolute_path, "wb")
        file.write(file_content)
        file.close()

    @staticmethod
    def rename_file(src_path, dst_path):
        """
        Moves a file in the filesystem, creates the directories needed
        :param src_path: Relative path of the file
        :param dst_path: Relative path where to put the file
        """
        src_path = StorageManager.get_absolute_path(src_path)
        dst_path = StorageManager.get_absolute_path(dst_path)
        StorageManager._create_dir(dst_path)
        os.rename(src_path, dst_path)

    @staticmethod
    def get_absolute_path(relative_path):
        """
        Get the absolute path of a stored file
        :param relative_path: Relative path of the file 
        :return: The absolute path of the file 
        """
        return os.path.abspath(os.path.join(Config.storedir, relative_path))

    @staticmethod
    def _create_dir(filename):
        """
        Create a directory in the filesystem
        :param filename: Absolute path of the file or its directory
        """
        dirname = os.path.dirname(filename)
        os.makedirs(dirname, exist_ok=True)

    @staticmethod
    def _sanitize(filename):
        # TODO vvvvvvvv
        return filename

File: backend/src/storage_manager.rs
use std::io;
use std::path::{Path, PathBuf};

use color_eyre::eyre::{Result, eyre};
use once_cell::sync::Lazy;
use regex::Regex;
use tokio::fs;

use crate::config;

const MAX_LENGTH: usize = 100;

/// Get the relative path to the source file with the specified filename
pub fn new_source_file(source_id: &str, filename: &str) -> Result<PathBuf> {
    let sanitized_filename = sanitize(filename)?;
    let mut path = PathBuf::from("source");
    path.push(&source_id[0..2]);
    path.push(&source_id[2..4]);
    path.push(source_id);
    path.push(sanitized_filename);
    Ok(path)
}

/// Get the relative path to the output file with the specified filename
pub fn new_output_file(output_id: &str, filename: &str) -> Result<PathBuf> {
    let sanitized_filename = sanitize(filename)?;
    let mut path = PathBuf::from("output");
    path.push(&output_id[0..2]);
    path.push(&output_id[2..4]);
    path.push(output_id);
    path.push(sanitized_filename);
    Ok(path)
}

/// Get the relative path to the input file for the specified attempt on the specified task
pub async fn new_input_file(input_id: &str, task_name: &str, attempt: u32) -> io::Result<PathBuf> {
    let filename = format!("{}_input_{}.txt", task_name, attempt);
    let mut path = PathBuf::from("input");
    path.push(&input_id[0..2]);
    path.push(&input_id[2..4]);
    path.push(input_id);
    path.push(filename);
    create_dir(&get_absolute_path(&path)).await?;
    Ok(path)
}

/// Get the size of the filename in bytes
pub async fn get_file_size(relative_path: &Path) -> io::Result<u64> {
    let absolute_path = get_absolute_path(relative_path);
    let metadata = fs::metadata(absolute_path).await?;
    Ok(metadata.len())
}

/// Store a file in the filesystem, creates the directories needed
pub async fn save_file(relative_path: &Path, file_content: &[u8]) -> io::Result<()> {
    let absolute_path = get_absolute_path(relative_path);
    create_dir(&absolute_path).await?;
    fs::write(absolute_path, file_content).await
}

/// Moves a file in the filesystem, creates the directories needed
pub async fn rename_file(src_path: &Path, dst_path: &Path) -> io::Result<()> {
    let src_abs = get_absolute_path(src_path);
    let dst_abs = get_absolute_path(dst_path);
    create_dir(&dst_abs).await?;
    fs::rename(src_abs, dst_abs).await
}

/// Get the absolute path of a stored file
pub fn get_absolute_path(relative_path: &Path) -> PathBuf {
    if relative_path.is_absolute() {
        return relative_path.to_path_buf();
    }
    Path::new(config::STORE_DIR).join(relative_path)
}

/// Create a directory in the filesystem
async fn create_dir(filename: &Path) -> io::Result<()> {
    if let Some(dirname) = filename.parent() {
        fs::create_dir_all(dirname).await
    } else {
        Ok(())
    }
}

/// Sanitize a filename.
fn sanitize(filename: &str) -> Result<String> {
    static RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"[^-a-zA-Z0-9_.]").unwrap());
    let filename = RE
        .replace_all(&filename.trim().replace(' ', "_"), "")
        .to_string();

    if filename.is_empty() || filename == "." || filename == ".." {
        return Err(eyre!("Invalid file name"));
    }

    let path = Path::new(&filename);
    let name = path
        .file_stem()
        .unwrap_or_default()
        .to_str()
        .unwrap_or_default();
    let ext = path
        .extension()
        .unwrap_or_default()
        .to_str()
        .unwrap_or_default();

    if ext.len() > MAX_LENGTH - 1 {
        return Ok(filename.chars().take(MAX_LENGTH).collect());
    }

    let name_len = std::cmp::min(
        name.len(),
        MAX_LENGTH - ext.len() - if ext.is_empty() { 0 } else { 1 },
    );
    let name = &name[0..name_len];

    if ext.is_empty() {
        Ok(name.to_string())
    } else {
        Ok(format!("{}.{}", name, ext))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize() {
        assert_eq!(sanitize("  file name.txt  ").unwrap(), "file_name.txt");
        assert_eq!(sanitize("file/name!@#$.txt").unwrap(), "filename.txt");
        assert_eq!(sanitize("a..b").unwrap(), "a..b");
        assert!(sanitize("").is_err());
        assert!(sanitize(".").is_err());
        assert!(sanitize("..").is_err());
        assert!(sanitize("  ").is_err());

        let long_name = "a".repeat(120);
        assert_eq!(sanitize(&long_name).unwrap().len(), MAX_LENGTH);

        let long_name_ext = format!("{}.txt", "a".repeat(120));
        let sanitized = sanitize(&long_name_ext).unwrap();
        assert_eq!(sanitized.len(), MAX_LENGTH);
        assert!(sanitized.ends_with(".txt"));
        assert_eq!(&sanitized[..95], &"a".repeat(95));
    }

    #[test]
    fn test_path_generation() {
        let source_path = new_source_file("12345678", "file.txt").unwrap();
        assert_eq!(source_path, PathBuf::from("source/12/34/12345678/file.txt"));

        let output_path = new_output_file("abcdefgh", "another file.zip").unwrap();
        assert_eq!(
            output_path,
            PathBuf::from("output/ab/cd/abcdefgh/another_file.zip")
        );
    }
}
